iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/project/tara/TaraReferenceExtractionUtils.java

```diff
@@ -0,0 +1,141 @@
+        Column joinExprs = documentMetadataWithHashDF.col("hashValue").equalTo(
+                documentHashFromCacheDF.col("hashValue"));
+        return documentMetadataWithHashDF
+                .join(documentHashFromCacheDF, joinExprs, "left_anti")
+                .select(
+                        col("id"),
+                        col("title"),
+                        col("abstract"),
+                        col("text")
+                );
+    }
+
+    private static final StructType PIPE_RESULT_SCHEMA = StructType$.MODULE$.apply(
+            Collections.singletonList(
+                    StructField$.MODULE$.apply("value", DataTypes.StringType, true, Metadata.empty())
+            )
+    );
+
+    public static Dataset<Row> runReferenceExtraction(SparkSession spark,
+                                                      Dataset<Row> documentMetadataDF,
+                                                      PipeExecutionEnvironment environment) throws IOException {
+        logger.info("Running reference extraction for input document metadata.");
+        String pipeCommandStr = environment.pipeCommand();
+        JavaRDD<Row> piped = documentMetadataDF.toJSON().javaRDD().pipe(pipeCommandStr).map(RowFactory::create);
+        return spark.createDataFrame(piped, PIPE_RESULT_SCHEMA)
+                .withColumn("json_struct", from_json(col("value"),
+                        (StructType) SchemaConverters.toSqlType(DocumentToProject.SCHEMA$).dataType()))
+                .select(expr("json_struct.*"));
+    }
+
+    public static Dataset<Row> documentHashToProjectToBeCached(SparkSession spark,
+                                                               Dataset<Row> documentToProjectDF,
+                                                               Dataset<Row> documentHashToProjectFromCacheDF,
+                                                               Dataset<Row> documentMetadataWithHashDF) {
+        logger.info("Finding reference extraction results to be cached.");
+        Column joinExprs = documentToProjectDF.col("documentId").equalTo(
+                documentMetadataWithHashDF.col("id"));
+        Dataset<Row> documentHashToProjectDF = documentToProjectDF
+                .join(documentMetadataWithHashDF, joinExprs)
+                .select(
+                        col("hashValue"),
+                        col("projectId"),
+                        col("confidenceLevel"),
+                        col("textsnippet")
+                );
+        Dataset<Row> toBeCached = documentHashToProjectFromCacheDF.union(documentHashToProjectDF);
+        return dataFrameWithSchema(spark, toBeCached, DocumentHashToProject.SCHEMA$);
+    }
+
+    public static Dataset<Row> documentHashToBeCached(SparkSession spark,
+                                                      Dataset<Row> documentHashFromCacheDF,
+                                                      Dataset<Row> documentMetadataWithHashDF) {
```

Comment:
We're not unioning with the whole `documentMetadataWithHashDF` here. We select only `hashValue` from this dataframe before unioning and the actual union is done for two dataframes with a single column containing hash value.

`distinct` is here not because of duplicates in the results of madis script execution (the results of reference extraction are not in the input parameters to this method). `distinct` is needed because this method takes as input a dataframe with document metadata with hash of the entire job input and a dataframe with document hash values from cache. The first dataframe will contain records that are also in the second dataframe, that is, records that were previously processed and are stored in cache. This method produces a dataframe with 'new' hash value set of the processed input document metadata.